app/src/main/java/com/rkm/tasky/repository/implementation/TaskyEventRepositoryImpl.kt

```diff
@@ -0,0 +1,115 @@
+import com.rkm.tasky.repository.mapper.toEvent
+import com.rkm.tasky.util.result.EmptyResult
+import com.rkm.tasky.util.result.Result
+import com.rkm.tasky.util.result.asEmptyDataResult
+import com.rkm.tasky.util.result.onFailure
+import com.rkm.tasky.util.result.onSuccess
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class TaskyEventRepositoryImpl @Inject constructor(
+    private val remoteDataSource: TaskyEventRemoteDataSource,
+    private val localDataSource: EventDao,
+    private val attendeeLocalDataSource: AttendeeDao,
+    private val photoLocalDataSource: PhotoDao,
+    private val syncDataSource: SyncDao,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+) : TaskyEventRepository {
+    override suspend fun getEvent(id: String): Result<Event, DatabaseError.ItemError> =
+        withContext(dispatcher) {
+            return@withContext when (val results = localDataSource.getAllEventDetails(id)) {
+                null -> Result.Error(DatabaseError.ItemError.ITEM_DOES_NOT_EXIST)
+                else -> Result.Success(results.toEvent())
+            }
+        }
+
+    override suspend fun createEvent(event: Event): EmptyResult<NetworkError.APIError> =
+        withContext(dispatcher) {
+            val result = safeCall {
+                remoteDataSource.createEvent(
+                    event.asCreateEventRequest(),
+                    event.photosToUpload.asMultiPartBody()
+                )
+            }
+            result.onSuccess {
+                localDataSource.upsertEvent(it.asEventEntity())
+                attendeeLocalDataSource.upsertAttendees(it.attendees.map { attendee -> attendee.asAttendeeEntity() })
+                photoLocalDataSource.upsertPhotos(it.photos.map { photo -> photo.asPhotoEntity(it.id) })
+            }
+
+            result.onFailure {
+                syncDataSource.upsertSyncItem(
+                    SyncEntity(
+                        action = SyncUserAction.CREATE,
+                        item = SyncItemType.EVENT,
+                        itemId = event.id
+                    )
+                )
+            }
+
+            return@withContext result.asEmptyDataResult()
+        }
+
+    override suspend fun updateEvent(event: Event): EmptyResult<NetworkError.APIError> =
+        withContext(dispatcher) {
+            val result = safeCall {
+                remoteDataSource.updateEvent(
+                    event.asUpdateEventRequest(),
+                    event.photosToUpload.asMultiPartBody()
+                )
+            }
+            result.onSuccess {
+                localDataSource.upsertEvent(it.asEventEntity())
+                attendeeLocalDataSource.upsertAttendees(it.attendees.map { attendee -> attendee.asAttendeeEntity() })
+                photoLocalDataSource.upsertPhotos(it.photos.map { photo -> photo.asPhotoEntity(it.id) })
```

Comment:
Same here, you wouldn’t want the coroutine to get cancelled and then not call all 3 of these functions. To protect against this, use `applicationScope.launch {}`